pkg/volume/csi/csi_attacher.go

```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
 }
 
 func (c *csiAttacher) GetDeviceMountPath(spec *volume.Spec) (string, error) {
-	glog.V(4).Info(log("attacher.GetDeviceMountPath is not implemented"))
-	return "", nil
+	glog.V(4).Info(log("attacher.GetDeviceMountPath(%v)", spec))
+
+	if spec == nil {
+		return "", fmt.Errorf("attacher.GetDeviceMountPath failed, spec is nil")
+	}
+
+	csiSource, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("attacher.GetDeviceMountPath failed getting CSI Source from Spec: %v", err))
+		return "", err
+	}
+	if csiSource.VolumeHandle == "" {
+		return "", fmt.Errorf("attacher.GetDeviceMountPath failed, volume handle empty")
+	}
+	if csiSource.Driver == "" {
+		return "", fmt.Errorf("attacher.GetDeviceMountPath failed, driver name empty")
+	}
+
+	deviceMountPath := path.Join(c.plugin.host.GetPluginDir(csiSource.Driver), "mounts", csiSource.VolumeHandle)
+	glog.V(4).Infof("attacher.GetDeviceMountPath succeeded, deviceMountPath: %s", deviceMountPath)
+	return deviceMountPath, nil
 }
 
 func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) error {
-	glog.V(4).Info(log("attacher.MountDevice is not implemented"))
+	glog.V(4).Infof(log("attacher.MountDevice(%s, %s)", devicePath, deviceMountPath))
+
+	// Setup
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	if spec == nil {
+		return fmt.Errorf("attacher.MountDevice failed, spec is nil")
+	}
+	csiSource, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("attacher.MountDevice failed to get CSI persistent source: %v", err))
+		return err
+	}
+
+	if c.csiClient == nil {
+		if csiSource.Driver == "" {
+			return fmt.Errorf("attacher.MountDevice failed, driver name is empty")
+		}
+		addr := fmt.Sprintf(csiAddrTemplate, csiSource.Driver)
+		c.csiClient = newCsiDriverClient("unix", addr)
+	}
+	csi := c.csiClient
+
+	// Check whether "STAGE_UNSTAGE_VOLUME" is set
+	capabilities, err := csi.NodeGetCapabilities(ctx)
+	if err != nil {
+		glog.Error(log("attacher.MountDevice failed to get node capabilities: %v", err))
+		return fmt.Errorf("attacher.MountDevice failed while getting node capabilities: %v", err)
+	}
+
+	if capabilities == nil {
+		glog.Infof(log("attacher.MountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping MountDevice..."))
+		return nil
+	}
+
+	stageUnstageSet, err := hasStageUnstageCapability(capabilities)
+	if err != nil {
+		glog.Error(log("attacher.MountDevice failed to check STAGE_UNSTAGE_VOLUME: %v", err))
+		return err
+	}
+	if !stageUnstageSet {
+		glog.Infof(log("attacher.MountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping MountDevice..."))
+		return nil
+	}
+
+	// Start MountDevice
+	if deviceMountPath == "" {
+		return fmt.Errorf("attacher.MountDevice failed, deviceMountPath is empty")
+	}
+
+	mounted, err := isDirMounted(c.plugin, deviceMountPath)
```

Comment:
Could you put this check first, and if it is false, then create a context, client, etc.?